// Controllers/calendar.js
var fileWriter = require('./fileWriter');

//---------- enumerators ---------------

var days = { "monday": [], "tuesday": [], "wednesday": [], "thursday": [], "friday": [], "saturday": [], "sunday": [] };
var outputFilename = "/Termostato/store/calendar.js";
var temps = [];

//---------- globals ---------------

var myCalendar = fileWriter.read(outputFilename);

//----------- initiator ------------

(function initCalendar() {
    if (myCalendar == undefined) { //file inesistente
        myCalendar = days;
        fileWriter.write(outputFilename, myCalendar);
    }
})();//initCalendar

//---------- privates ---------------

//---------- methods ---------------

var add = function(data){
	
	var timeSpan = {
						start: data.startTime, 
						end: data.endTime, 
						temperature: data.temp
	};

	if (timeSpan.start != undefined && timeSpan.end != undefined) { //devo avere le date impostate senno non procedo

	    switch (data.day) {
	        case 'monday':
	            myCalendar.monday.push(timeSpan);
	            break;
	        case 'tuesday':
	            myCalendar.tuesday.push(timeSpan);
	            break;
	        case 'wednesday':
	            myCalendar.wednesday.push(timeSpan);
	            break;
	        case 'thursday':
	            myCalendar.thursday.push(timeSpan);
	            break;
	        case 'friday':
	            myCalendar.friday.push(timeSpan);
	            break;
	        case 'saturday':
	            myCalendar.saturday.push(timeSpan);
	            break;
	        case 'sunday':
	            myCalendar.sunday.push(timeSpan);
	            break;
	    }
	    fileWriter.write(outputFilename, myCalendar);
	    return { 'day': data.day, 'calendar': myCalendar[data.day] };
	}
};//add

var remove = function(data){
	
	switch (data.day){
		case 'monday':
				myCalendar.monday.splice(data.pos,1);
		break;
		case 'tuesday':
				myCalendar.tuesday.splice(data.pos,1);
		break;
		case 'wednesday':
				myCalendar.wednesday.splice(data.pos,1);
		break;
		case 'thursday':
				myCalendar.thursday.splice(data.pos,1);
		break;
		case 'friday':
				myCalendar.friday.splice(data.pos,1);
		break;
		case 'saturday':
				myCalendar.saturday.splice(data.pos,1);
		break;
		case 'sunday':
				myCalendar.sunday.splice(data.pos,1);
		break;
	}//switch
	fileWriter.write(outputFilename, myCalendar);
	return { 'day': data.day, 'calendar': myCalendar[data.day] };
};//remove

var removeAll = function(){
	myCalendar.monday = [];
	myCalendar.tuesday = [];
	myCalendar.wednesday = [];
	myCalendar.thursday = [];
	myCalendar.friday = [];
	myCalendar.sunday = [];
	myCalendar.saturday = [];
	fileWriter.write(outputFilename, myCalendar);
};//removeAll

(function () {
    if (myCalendar == {})
    {
        removeAll();
    }
})();//le () chimano la funzione anonima

function checkSchedule(callback, temperature) {
    if (temps.length < 5)
    {
        temps.push(temperature);
        return;
    }
    else
    {
        //calcola la media di 5 temperature misurate
        var sum = temps.reduce(function(a, b) { return a + b; });
        temperature = (sum / temps.length).toFixed(1);
        console.log(temps);
        temps = [];
    }

	var currentDate = new Date();
	var current = (currentDate.getHours() * 60) + currentDate.getMinutes();
	var isOn = false;

	if(currentDate.getDay() == 6)//saturday
	{
	    isOn = check(myCalendar.saturday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 5)//saturday
	{
	    isOn = check(myCalendar.friday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 4)//saturday
	{
	    isOn = check(myCalendar.thursday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 3)//saturday
	{
	    isOn = check(myCalendar.wednesday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 2)//saturday
	{
	    isOn = check(myCalendar.tuesday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 1)//saturday
	{
	    isOn = check(myCalendar.monday, current, temperature);
	}//if
	
	if(currentDate.getDay() == 0)//saturday
	{
	    isOn = check(myCalendar.sunday, current, temperature);
	}//if

	if (isOn) { callback.setOn(); }
	else { callback.setOff(); }
}//checkSchedule

function check(calendarDay, current, temperature)
{
	for(i=0;i<calendarDay.length;i++)
	{
	    var start = (parseInt(calendarDay[i].start.split(':')[0], 10) * 60) + parseInt(calendarDay[i].start.split(':')[1], 10);
	    //console.log(i + '  ' + calendarDay.length + "  " + start + " " + current + '   ' + "Termperature: " + temperature + " <= " + calendarDay[i].temperature +' ?' );
		if(start <= current)
		{
		    var end = (parseInt(calendarDay[i].end.split(':')[0], 10) * 60) + parseInt(calendarDay[i].end.split(':')[1], 10);
		    
		    if (end >= current) {
		        console.log("Termperature: " + temperature + " <= " + calendarDay[i].temperature + ' ?');
		        var isOn = temperature <= calendarDay[i].temperature;
		        return isOn; //temperature <= calendarDay[i].temperature ? true : false;
		    }//if
		}
	}//for
	return false;
}

//---------- exports ---------------

exports.myCalendar = myCalendar;
exports.add = add;
exports.remove = remove;
exports.removeAll = removeAll;
exports.checkSchedule = checkSchedule;
