// Controllers/calendar.js
var fileWriter = require('./fileWriter');

//---------- enumerators ---------------

var days = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"];
var outputFilename = "/Termostato/store/calendar.js";

//---------- globals ---------------

var myCalendar = fileWriter.read( outputFilename );

var startUpScheduleChecker = setInterval(function(){ checkSchedule(); }, 3000);

//---------- privates ---------------

//---------- methods ---------------

var add = function(data){
	
	var timeSpan = {
						start: data.startTime, 
						end: data.endTime, 
						temperature: data.temp
					};
	
	switch (data.day){
		case 'monday':
			myCalendar.monday.push(timeSpan);
		break;
		case 'tuesday':
			myCalendar.tuesday.push(timeSpan);
		break;
		case 'wednesday':
			myCalendar.wednesday.push(timeSpan);
		break;
		case 'thursday':
			myCalendar.thursday.push(timeSpan);
		break;
		case 'friday':
			myCalendar.friday.push(timeSpan);
		break;
		case 'saturday':
			myCalendar.saturday.push(timeSpan);
		break;
		case 'sunday':
			myCalendar.sunday.push(timeSpan);
		break;
	}
	fileWriter.write(outputFilename, myCalendar);
	return { 'day': data.day, 'calendar': myCalendar[data.day] };
};//add

var remove = function(data){
	
	switch (data.day){
		case 'monday':
				myCalendar.monday.splice(data.pos,1);
		break;
		case 'tuesday':
				myCalendar.tuesday.splice(data.pos,1);
		break;
		case 'wednesday':
				myCalendar.wednesday.splice(data.pos,1);
		break;
		case 'thursday':
				myCalendar.thursday.splice(data.pos,1);
		break;
		case 'friday':
				myCalendar.friday.splice(data.pos,1);
		break;
		case 'saturday':
				myCalendar.saturday.splice(data.pos,1);
		break;
		case 'sunday':
				myCalendar.sunday.splice(data.pos,1);
		break;
	}//switch
	fileWriter.write(outputFilename, myCalendar);
	return { 'day': data.day, 'calendar': myCalendar[data.day] };
};//remove

var removeAll = function(){
	myCalendar.monday = [];
	myCalendar.tuesday = [];
	myCalendar.wednesday = [];
	myCalendar.thursday = [];
	myCalendar.friday = [];
	myCalendar.sunday = [];
	myCalendar.saturday = [];
	fileWriter.write(outputFilename, myCalendar);
};//removeAll

function checkSchedule(){
	var currentDate = new Date();
	var current = parseInt( (""+currentDate.getHours()) + (""+currentDate.getMinutes() ), 10);

	if(currentDate.getDay() == 6)//saturday
	{
		check(myCalendar.saturday, current);
	}//if
	
	if(currentDate.getDay() == 5)//saturday
	{
		check(myCalendar.friday), current;
	}//if
	
	if(currentDate.getDay() == 4)//saturday
	{
		check(myCalendar.thursday, current);
	}//if
	
	if(currentDate.getDay() == 3)//saturday
	{
		check(myCalendar.wednesday, current);
	}//if
	
	if(currentDate.getDay() == 2)//saturday
	{
		check(myCalendar.tuesday, current);
	}//if
	
	if(currentDate.getDay() == 1)//saturday
	{
		check(myCalendar.monday, current);
	}//if
	
	if(currentDate.getDay() == 0)//saturday
	{
		check(myCalendar.sunday, current);
	}//if
}//checkSchedule

function check(calendarDay, current)
{
	for(i=0;i<calendarDay.length;i++)
	{
		var start = parseInt(calendarDay[i].start.split(':')[0] + calendarDay[i].start.split(':')[1], 10);
		if(start <= current) 
		{
			var end = parseInt(calendarDay[i].end.split(':')[0] + calendarDay[i].end.split(':')[1], 10);
			if(end >= current) 
			{
				console.log("Ho impostato la temperatura a " + calendarDay[i].temperature);
			}
		}
	}//for
}

//---------- exports ---------------

exports.myCalendar = myCalendar;
exports.add = add;
exports.remove = remove;
exports.removeAll = removeAll;
