// Controllers/fileWriter.js
var fs = require('fs');

var write = function(outputFilename, dataObject) {
    fs.writeFile(outputFilename, JSON.stringify(dataObject, null, 4), function (err) {
        if (err) {
            console.log(err);
        } else {
            console.log("JSON saved to " + outputFilename);
        }
    });
}//write

var append = function (outputFilename, dataObject) {
    fs.appendFile(outputFilename, dataObject, function (err) {
        if (err) {
            console.log(err);
        } else {
            console.log("Appended to file " + outputFilename);
        }
    });
}//write

//Sync passerà il valore alla varibile - ho provato a passare il riferimento e il metodo non Sync e non funziona?!?!
var read = function(outputFilename){ 
	return JSON.parse( fs.readFileSync(outputFilename, 'utf8') ); //read
}

//Sync passerà il valore alla varibile - ho provato a passare il riferimento e il metodo non Sync e non funziona?!?!
var readText = function (outputFilename) {
    try
    {
        return fs.readFileSync(outputFilename, 'utf8'); //read
    }
    catch (e)
    {
        if (e.code === 'ENOENT') {
            console.log('File not found!');
        } else {
            throw e;
        }
    }
    return "";
}

//Sync passerà il valore alla varibile - ho provato a passare il riferimento e il metodo non Sync e non funziona?!?!
var appendLog = function (basePath, fileName, date, dataObject) {
    var day = setZeros(date.getDate());
    var month = setZeros( (date.getMonth() + 1) );

    var directory = basePath + '/' + date.getFullYear() + '-' + month + '-' + day;
    if (!fs.existsSync(directory)) {
        fs.mkdirSync(directory);
    }
    append(directory + '/' + fileName, dataObject);
}//appendLog

function setZeros(val)
{
    return val.lenght > 1 ? val : '0' + val;
}//setZeros

//---------- exports ---------------

exports.append = append;
exports.write = write;
exports.read = read;
exports.readText = readText;
exports.appendLog = appendLog;
